source/base/parameter_handler.cc

```diff
@@ -1773,59 +1762,45 @@ namespace
-                              << "        " << new_value << std::endl
-                              << "    for the entry named" << std::endl
-                              << "        " << full_path << std::endl
-                              << "    does not match the given pattern" << std::endl
-                              << "        " << patterns[pattern_index]->description()
-                              << std::endl;
-                    return false;
-                  }
-
-                // set the found parameter in
-                // the destination argument
-                destination.put (full_path + path_separator + "value",
-                                 new_value);
-
-                // this node might have
-                // sub-nodes in addition to
-                // "value", such as
-                // "default_value",
-                // "documentation", etc. we
-                // might at some point in the
-                // future want to make sure
-                // that if they exist that
-                // they match the ones in the
-                // 'destination' tree
-              }
-            else
-              {
-                std::cerr << "The entry <" << full_path
-                          << "> with value <"
-                          << p->second.get<std::string>("value")
-                          << "> has not been declared."
-                          << std::endl;
-                return false;
-              }
+
+            const std::string new_value
+              = p->second.get<std::string>("value");
+            AssertThrow (destination.get_optional<std::string> (full_path)
+                         &&
+                         destination.get_optional<std::string>
+                         (full_path + path_separator + "value"),
+                         ParameterHandler::ExcEntryUndeclared (p->first));
+
+            // first make sure that the
+            // new entry actually
+            // satisfies its constraints
+            const unsigned int pattern_index
+              = destination.get<unsigned int> (full_path +
+                                               path_separator +
+                                               "pattern");
+            AssertThrow (patterns[pattern_index]->match(new_value),
+                         ParameterHandler::ExcInvalidEntryForPatternXML
+                         // XML entries sometimes have extra surrounding
+                         // newlines
+                         (Utilities::trim(new_value),
+                          p->first,
+                          patterns[pattern_index]->description()));
+
+            // set the found parameter in
+            // the destination argument
+            destination.put (full_path + path_separator + "value",
+                             new_value);
+
+            // this node might have
+            // sub-nodes in addition to
```

Comment:
while you're there, can you rewrap this comment so it doesn't take up all of these lines?